### ISSUE TITLE
Gradient Preconditioner

### ISSUE THREAD

**Issue by shannon63** [google/jax#2197]
I'm new to JAX and am trying to implement preconditioned SGD. I was wondering if there is a recommended way of doing this using optimizers.sgd()? For example, in the following code snippet (taken from https://github.com/google/jax/blob/master/docs/notebooks/maml.ipynb)

```
def step(i, opt_state, x1, y1):
  p = get_params(opt_state)
  g = grad(loss)(p, x1, y1)
  return opt_update(i, g, opt_state)
```

should I just pass P*g instead of g, where P is some preconditioner matrix? If so, given that g will be some tuple, what format datatype should P be in to make the matrix-vector multiplication possible? Any pointers would be appreciated!

Thank you!

**Comment by shoyer** [comment]
I would suggest taking a look at the source code for the optimizers module. It is pretty simple/lightweight.

**Comment by mattjj** [comment]
Adding on to what @shoyer said, I'd say don't feel constrained by the optimizers package as an API! Just open it up and write the code however you like. Most of the functions in optimizers.py implement some kind of preconditioned SGD, usually with diagonal preconditioning, and they're just plain-old NumPy expressions. You can probably follow a similar pattern.

Hope that answers your question! Please reopen, or open a new one, if needed.

**Comment by shannon63** [comment]
Thanks @shoyer and @mattjj-- I'm sorry for the slow reply!

I was hoping to implement natural gradient descent or Gauss-Newton, for which I have to invert a large p-by-p matrix, where p is the number of parameters of my neural network. I understand that if I only want to perform a small number of gradient steps, I could implement H^-1 * g by using conjugate gradient descent, but I'm not sure how to best do that in JAX. Is there an existing implementation of CG in JAX? I haven't been able to find one. In general, am I right to assume that using the full inverse Hessian as a preconditioner should be feasible with CG in JAX?

Thank you!
